### PR TITLE
[Snippets] Introduced helpers insert_node and replace_node

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Removed `replace_input` from LinearIR since it's not responsibility of LinearIR and added this support to `Expression` and `ExpressionPort`*
 - *Added helpers `replace_with_expr`, `replace_with_node` and `insert_node` to LinearIR to encapsulate common work with expression creation, insertion etc. Fore more details, please read briefs of the helpers in LinearIR*

### Tickets:
 - *122690, 126994*

### CODE REVIEW

**Review by v-Golubev — COMMENTED** [review]
Only minor suggestions for now

File: src/common/snippets/src/lowered/loop_manager.cpp [hunk 1]
```diff
@@ -188,12 +188,24 @@ LoopInfoPtr LinearIR::LoopManager::get_loop_info(size_t index) const {
...
 
+std::vector<size_t> LinearIR::LoopManager::get_common_outer_loops(const ExpressionPtr& lhs, const ExpressionPtr& rhs) {
+    const auto& rhs_ids = rhs->get_loop_ids();
+    const auto& lhs_ids = lhs->get_loop_ids();
+    size_t idx = 0;
+    while (idx < std::min(rhs_ids.size(), lhs_ids.size())) {
+        if (rhs_ids[idx] != lhs_ids[idx])
+            break;
+        idx++;
+    }
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/lowered/loop_manager.cpp]: Right! Thanks for the idea

File: src/common/snippets/src/op/subgraph.cpp [hunk 2]
```diff
@@ -42,6 +42,7 @@
 #include "snippets/lowered/pass/insert_loops.hpp"
 #include "snippets/lowered/pass/optimize_domain.hpp"
 #include "snippets/lowered/pass/insert_perf_count.hpp"
+#include "snippets/lowered/pass/serialize_control_flow.hpp"
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/op/subgraph.cpp]: Left after debugging. Removed, thanks!

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 3]
```diff
@@ -314,6 +290,110 @@ VectorDims LinearIR::get_master_shape() const {
...
+LinearIR::exprIt LinearIR::insert_node(const std::shared_ptr<ov::Node>& new_node, const std::vector<size_t>& loop_ids, bool update_loop_ports,
+                                       const constExprIt& place, const std::set<ExpressionPort>& consumers) {
+    const auto consumers_py_port = consumers.empty() ? std::vector<std::set<ExpressionPort>>{} : std::vector<std::set<ExpressionPort>>{ consumers };
+    return insert_node(new_node, loop_ids, update_loop_ports, place, consumers_py_port);
+}
+
+LinearIR::exprIt LinearIR::replace_node(const std::shared_ptr<ov::Node>& new_node, const std::vector<ExpressionPtr>& old_exprs,
+                                        const std::vector<size_t>& loop_ids, const constExprIt& place) {
+    OPENVINO_ASSERT(!old_exprs.empty(), "Failed to replace node: there are no old expressions for replacing");
+     OPENVINO_ASSERT(new_node->get_output_size() == old_exprs.back()->get_output_count(),
```
Comment (v-Golubev) [inline src/common/snippets/src/lowered/linear_ir.cpp:332]: Minor: extra indentation
```suggestion
    OPENVINO_ASSERT(new_node->get_output_size() == old_exprs.back()->get_output_count(),
```

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 4]
```diff
@@ -314,6 +290,110 @@ VectorDims LinearIR::get_master_shape() const {
...
+                consumers[port_connector->get_source().get_index()] = old_expr_consumers;
+            }
+        }
+    }
+    for (size_t i = 0; i < new_node->get_output_size(); ++i)
+        snippets::lowered::PortDescriptorUtils::set_port_descriptor_ptr(new_node->output(i), last_old_expr->get_output_port_descriptor(0)->clone());
+
+    // Notes:
+    // - we cannot automatically updated loop ports in node insertion since there are old_exprs
+    // - we should update loop ports firstly and onlyt after that - remove old exprs from linearIR
```
Comment (v-Golubev) [inline src/common/snippets/src/lowered/linear_ir.cpp:374]: Typo
```suggestion
    // Notes:
    // - we cannot automatically update loop ports in node insertion since there are old_exprs
    // - we should update loop ports firstly and only after that - remove old exprs from linearIR
```

**Review by v-Golubev — APPROVED** [review]
LGTM

File: src/common/snippets/src/lowered/linear_ir.cpp (hunk 3 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/linear_ir.cpp:332]: Sure, thanks

File: src/common/snippets/src/lowered/linear_ir.cpp (hunk 4 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/linear_ir.cpp:374]: Done, thanks

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 5]
```diff
@@ -139,6 +136,74 @@ class LinearIR {
     const std::shared_ptr<ShapeInferSnippetsNode>& get_shape_infer_instance() const {return m_shape_infer; }
     VectorDims get_master_shape() const;
 
+    /* ------ Helpers for work with LinearIR ----- */
```
Comment (IvanNovoselov) [inline src/common/snippets/include/snippets/lowered/linear_ir.hpp:139]: Regarding all the helpers: in general, I think that using ov::Node in LIR helpers is not a good practice because the Node can't fully describe the LIR data flow, and the validity of ov::model connections is not guaranteed at this point.
Although I agree that it is a tempting compromise for the `insert_node` methods (at least a user can provide connectors manually), I think it's worth to consider changing `replace_node` signatures to work with expressions only.
Please see more detailed comments in the corresponding methods.

File: src/common/snippets/src/lowered/pass/load_movebroadcast_to_broadcastload.cpp [hunk 6]
```diff
@@ -27,7 +26,7 @@ bool LoadMoveBroadcastToBroadcastLoad::run(LinearIR& linear_ir) {
             const auto& interm_connector = expr->get_input_port_connector(0);
             const auto parent_expr = interm_connector->get_source().get_expr();
             const auto load = ov::as_type_ptr<op::Load>(parent_expr->get_node());
-            if (!load)
+            if (!load || expr->get_loop_ids() != parent_expr->get_loop_ids())
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/load_movebroadcast_to_broadcastload.cpp:29]: Why do we change this transformation behavior in this PR? It there a reported bug or undesired behavior?
It is not clear why these changes are valid straight away.

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 7]
```diff
@@ -139,6 +136,74 @@ class LinearIR {
...
+     *           and all sources of the `old_exprs` must be expression from this seq as well or must be source of `new_inputs`
+     *        - The helper set output PortDescriptors - clones from last removable expression
+     *        - The helpers updates LoopPorts of the corresponding loops using information about removable expressions
+     * @param new_node the target node
+     * @param old_exprs the sequence of removable expressions
+     * @param loop_ids vector of loops ids that will be set for the expression
+     * @param place before this place expression will be inserted
+     * @return new expression iterator in LinearIR
+     */
+    exprIt replace_node(const std::shared_ptr<ov::Node>& new_node, const std::vector<ExpressionPtr>& old_exprs, const std::vector<size_t>& loop_ids,
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp]: I've understood why I see many comments regarding misunderstanding `replace_node` methods!
Firstly, I mixed up the order in `replacement` sentence (what do we replace and what do we replacing with)
Secondly, methods `replace_node` have incorrect naming. As I understand, should be `replace_with_node` - I've updated.
I'm sorry for the that :sweat_smile:

File: src/common/snippets/include/snippets/lowered/expression_port.hpp [hunk 8]
```diff
@@ -42,6 +42,10 @@ class ExpressionPort {
     //  - Output port returns all consumer ports (children)
     std::set<ExpressionPort> get_connected_ports() const;
 
+    // Note: It may be called only for input expression ports
+    //       since output ports don't support PortConnector changing (this is determined by the creation of the expression)
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/expression_port.hpp:46]: 1) I see your point, let's leave as is. Only I would rather move this comment inside the function, where the PortConnector type is checked. This is up to you.
2) Not regarding this PR, just to clarify. Output connectors are not imprinted in expression by design, they're stored there as a vector. They are imprinted there by our factory, and can not be changed due to lack of appropriate API. We didn't create such an api on purpose, but I think it's time to re-evaluate this decision.
For example, we might need to modify output connector whenever we need to keep corresponding register alive for tail processing purposes, or some weird accumulation/reduction patterns. We also might need to remove "useless" connectors that are not required due to specific backend implementation, PerfCounters for example.
So my main message here, is that we need to keep our mind open, and probably we do need to modify output connectors (even for only rather exotic cases).

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 9]
```diff
@@ -314,6 +290,110 @@ VectorDims LinearIR::get_master_shape() const {
...
+
+LinearIR::exprIt LinearIR::replace_node(const std::shared_ptr<ov::Node>& new_node, const std::vector<ExpressionPtr>& old_exprs,
+                                        const std::vector<size_t>& loop_ids, const constExprIt& place) {
+    OPENVINO_ASSERT(!old_exprs.empty(), "Failed to replace node: there are no old expressions for replacing");
+    OPENVINO_ASSERT(new_node->get_output_size() == old_exprs.back()->get_output_count(),
+                   "Failed to replace node: node output port count is not equal to output count of last old expression");
+
+    std::vector<PortConnectorPtr> new_inputs(new_node->get_input_size());
+    for (size_t i = 0; i < new_node->get_input_size(); ++i) {
+        const auto& source = new_node->get_input_source_output(i);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/linear_ir.cpp:343]: It's probably not the best example with `LoopBegin`, but you can make another one with `Buffer` and other service ops. [Here](https://github.com/openvinotoolkit/openvino/blob/6150572c0ac9e188b629a2a5a33fa8d4e1960d0f/src/common/snippets/src/lowered/pass/insert_buffers.cpp#L183) for example, we don't call `replace_node` for ov::model when the buffer is inserted, and we don't update inputs of parent's outputs manually.
The point is that we do not **guarantee** ov model validity after LIR creation, so we shouldn't rely on nodes connections. Yes, we often connect nodes to appropriate parents, but this is done mostly to avoid any issues with validate_and_infer_types when the node is created or cloned. The only valid use of ov::Node in LIR (and also in ideal world) is to describe operation semantics. In future, we will encapsulate operation semantics directly in expressions and will get rid of ov::Node entirely.
So I think the best long-term strategy for Snippets is to gradually eliminate usage of ov::Node. I see your point, and I understand why it works, but it doesn't mean we should use it 🥲 
Let's continue this discussion offline. I'm sure we can find a future-proof solution together 🙂

File: src/common/snippets/src/lowered/linear_ir.cpp [hunk 10]
```diff
@@ -314,6 +290,110 @@ VectorDims LinearIR::get_master_shape() const {
...
+            }
+        }
+    }
+    for (size_t i = 0; i < new_node->get_output_size(); ++i)
+        snippets::lowered::PortDescriptorUtils::set_port_descriptor_ptr(new_node->output(i), last_old_expr->get_output_port_descriptor(0)->clone());
+
+    // Notes:
+    // - we cannot automatically update loop ports in node insertion since there are old_exprs
+    // - we should update loop ports firstly and only after that - remove old exprs from linearIR
+    const auto new_expr_it = insert_node(new_node, new_inputs, loop_ids, false, place, consumers);
```
[the remark this answers is not in the document]
Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/linear_ir.cpp]: yeah, all right. I got confused with the old `replace_node` signature and description 😅

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp (hunk 5 above)
Comment (IvanNovoselov) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp:139]: Good!

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/include/snippets/lowered/linear_ir.hpp]: Good idea!
Thanks a lot for the suggestion! Applied

**Review by IvanNovoselov — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
